Learner profile page.

muzaffaryousaf opened the pull request:
This PR includes leaner profile page with all the views(read-only, edit, limited) that are mentioned in the acceptance criteria.

The jira tickets are:
https://openedx.atlassian.net/browse/TNL-1501
https://openedx.atlassian.net/browse/TNL-1502

The sandbox is available here: http://mzfr-profile-page.m.sandbox.edx.org/

muzaffaryousaf commented:
@explorerleslie @marcotuts @srpearce can you please review this PR. ?

muzaffaryousaf commented:
@symbolist here.

marcotuts commented on lms/static/sass/base/_variables.scss:
```diff
@@ -187,6 +187,7 @@ $m-blue-t0: rgba($m-blue,0.125);
 $m-blue-t1: rgba($m-blue,0.25);
 $m-blue-t2: rgba($m-blue,0.50);
 $m-blue-t3: rgba($m-blue,0.75);
+$m-blue-hover: rgb(223,242,247); // #DFF2F7
```
we should rely on an existing blue here instead of a separate hover blue if we can. Additionally, it would be helpful to call this out by separating it with a comment #used in learner profile styling (so that we can more easily clean up our blue color usage later)

muzaffaryousaf replied:
@marcotuts can you please suggest any existing colour that we can use for this hover state ?

marcotuts replied:
I would try $m-blue-l4 (that seems like the closest color)

symbolist commented on lms/static/js/student_account/views/account_settings_fields.js:
```diff
@@ -205,48 +276,204 @@
             },
 
+            modelValueIsSet: function() {
+                return (this.modelValue() && this.optionForValue(this.modelValue())) == true;
```
@muzaffaryousaf I think this should resolve the issue where dropdowns with values are showing the wrong hover styling.
```
var value = this.modelValue();
if (_.isUndefined(value) || _.isNull(value) || value == '') {
    return false;
} else {
    return !(_.isUndefined(this.optionForValue(value)))
}
```

explorerleslie commented:
@muzaffaryousaf looking great! A few nits: 

- When you switch to Limited Profile, the text "This edX learner is not currently sharing their profile information" looks like it's clickable (when you hover, the hand icon appears). We should remove that behavior if possible.
- Same thing as above for the label text for "edX learners can see my:" text. We should not make it clickable when clicking doesn't do anything. 

Also, it was so cool to log in as "honor" and go to http://mzfr-profile-page.m.sandbox.edx.org/u/staff to see another user's profile. :)

marcotuts commented:
@muzaffaryousaf  - This is looking great! One thing after reviewing this with @explorerleslie was a suggestion to make the background of this page white. (cc'ed @frrrances ) To do this, we've in the past used a <body> level class with the "view-" prefix. This lets us specify page-specific styling changes, such as the background color change to white / very light gray.

srpearce commented on lms/static/js/student_account/views/account_settings_fields.js:
```diff
@@ -0,0 +1,554 @@
+            message: function (message) {
+                if (this.profileIsPrivate) {
+                    this._super(interpolate_text(
+                        gettext('Your profile is set to limited because you have not set your birth year yet. Set it on the {account_settings_page_link}.'),
```
Text: Before you share your full profile, you must set your birth year. To set your birth year, go to the {account_settings_page_link}.

muzaffaryousaf replied:
@explorerleslie should we add the age limit message here in this PR, because right now, there is no such limit.

srpearce replied:
@explorerleslie Did we decide this morning that MVP would have the age limit? If so, do we want to let users know about it here, or should we go with the age-less text?

srpearce commented on lms/templates/student_profile/learner_profile.underscore:
```diff
@@ -0,0 +1,28 @@
+            <div class="profile-section-two-fields">
+                <% if (!showFullProfile) { %>
+                    <label class="profile-private--message">
+                        <%- gettext('This edX learner is not currently sharing their profile details.') %>
```
Text: This edX learner is currently sharing a limited profile.

Comment: Should this message be different for the user herself and for other page viewers? It's a little strange for me to see "This edX learner is not currently sharing their profile details" on my own profile page. If we don't want to have different messages, can we indicate somewhere on the UI that other users will see the profile page exactly as it looks here - maybe a "Other edX learners will see your profile exactly as it appears below" under the "edX learners can see my" line? (Maybe this is a follow-on story?)

Another potential follow-on story: Let users know that if they make changes to the full profile, those changes won't be deleted if the user switches to the limited profile and then back to the full profile. It could be a message like "This information will be saved but not visible when you switch to a limited profile" (though said much more elegantly than this) above the "About Me" text box. (I would add this above the text box rather than as placeholder text so that it's always visible.)

muzaffaryousaf replied:
@explorerleslie @frrrances what do you think about these suggestions.

srpearce commented on lms/static/js/student_profile/views/learner_profile_factory.js:
```diff
@@ -0,0 +1,123 @@
+                    editable: editable,
+                    showMessages: false,
+                    title: gettext('About me'),
+                    placeholderValue: gettext("Tell other edX learners a little about yourself, where you're from, what your interests are, why you joined edX, what you hope to learn..."),
```
Tell other edX learners a little about yourself: where you live, what your interests are, why you’re taking courses on edX, or what you hope to learn.

srpearce replied:
Does this field have a character limit?

muzaffaryousaf replied:
Not really, It's just a placeholder for textarea.

srpearce commented:
@muzaffaryousaf @explorerleslie @marcotuts Sorry, all, I changed the email address for the staff@example.com account on the Account Settings page to my edX address (spearce@edx.org) before I realized I wouldn't be able to change it back. My bad. :-(

srpearce commented on lms/templates/navigation-edx.html:
```diff
@@ -79,9 +79,8 @@ <h1 class="logo" itemscope="" itemtype="http://schema.org/Organization">
+            <li><a href="${reverse('account_settings')}">${_("Account Settings")}</a></li>
+            <li><a href="${reverse('learner_profile', kwargs={'username': user.username})}">${_("My Profile")}</a></li>
           </%block>
           <li><a href="${reverse('logout')}" role="menuitem">${_("Sign out")}</a></li>
```
I know this isn't part of this PR, but can we change the text here to "Sign Out" so that the capitalization is consistent for all three options?

srpearce commented on lms/templates/navigation.html:
```diff
@@ -79,9 +79,8 @@ <h1 class="logo">
+            <li><a href="${reverse('account_settings')}">${_("Account Settings")}</a></li>
+            <li><a href="${reverse('learner_profile', kwargs={'username': user.username})}">${_("My Profile")}</a></li>
           </%block>
           <li><a href="${reverse('logout')}" role="menuitem">${_("Sign out")}</a></li>
```
I know this isn't part of this PR, but can we change the text here to "Sign Out" so that the capitalization is consistent for all three options?

frrrances commented:
@muzaffaryousaf Looking good! I have done a visual review and have some small things I will note so you can start working on them: 

- profile content area background-color should be white (full width from the visibility bar to the footer)
- hover icon color should be link blue ($link-blue)
- add a 3px border radius on all fields so hover is rounded
- placeholder text (“Add a language” “Add a location”) should appear when the field is empty/not set
- border on empty field on hover is not aligned right
![screen shot 2015-03-25 at 7 24 49 am](https://cloud.githubusercontent.com/assets/4327102/6823595/16b1119e-d2c0-11e4-8cc1-850df6e87ed0.png)

- increase padding at bottom of page
- user name should be bold (font-weight: 600) - looks like an input style is overriding the span where you have it set)
- location and language should be a mid-grey ($gray-l1) and 14px (use font-mixin or %t-copy-sub1)
- empty field dashed border should be 2px and $gray-l3 for default, and $link-blue on hover 
- use font awesome icon for the plus sign in + About Me
- font-size for placeholder About me should be 14px and $gray-l1

I will also go through the code and give you feedback on that later today.

muzaffaryousaf commented:
@andy-armstrong me & @muhammad-ammar done with the code for this PR. Would you please review the changes. Thanks.

frrrances commented on lms/static/sass/shared/_fields.scss:
```diff
@@ -0,0 +1,102 @@
+    border: 1px dashed transparent;
+
+    &.mode-placeholder:hover {
+        border: 1px dashed $m-blue-d1;
```
make this 2px, and use $link-blue for the color.

muzaffaryousaf replied:
By `$link-blue` do you mean `$link-color` coz i'm not able to find variable with this name.

frrrances commented on lms/static/sass/shared/_fields.scss:
```diff
@@ -0,0 +1,102 @@
+.u-field {
+    padding: $baseline 0;
+    border-bottom: 1px solid $gray-l5;
+    border: 1px dashed transparent;
```
add 3px border-radius.

frrrances commented on lms/static/sass/shared/_fields.scss:
```diff
@@ -0,0 +1,102 @@
+.u-field-title {
+    width: flex-grid(3, 12);
+    display: inline-block;
+    color: $dark-gray1;
```
use $gray-l1

frrrances commented on lms/static/sass/shared/_fields.scss:
```diff
@@ -0,0 +1,102 @@
+    width: flex-grid(6, 12);
+    display: inline-block;
+    vertical-align: top;
+    color: $dark-gray1;
```
use $gray-l1

frrrances commented on lms/static/sass/shared/_fields.scss:
```diff
@@ -0,0 +1,102 @@
+}
+
+.u-field-message {
+    @extend small;
```
use %t-copy-sub1

frrrances commented on lms/static/sass/views/_learner-profile.scss:
```diff
@@ -0,0 +1,183 @@
+
+.ui-loading-indicator {
+    @extend .ui-loading-base;
+
```
remove extra line

frrrances commented on lms/static/sass/views/_learner-profile.scss:
```diff
@@ -0,0 +1,183 @@
+    // center horizontally
+    @include margin-left(auto);
+    @include margin-right(auto);
+    width: $baseline*5;
```
wrap math: 

```
width: ($baseline*5);
```

frrrances commented on lms/static/sass/views/_learner-profile.scss:
```diff
@@ -0,0 +1,183 @@
+}
+
+.wrapper-profile-section-one {
+
```
remove extra line

frrrances commented on lms/static/sass/views/_learner-profile.scss:
```diff
@@ -0,0 +1,183 @@
+}
+
+.profile-section-one-fields {
+
```
remove extra line

frrrances commented on lms/static/sass/views/_learner-profile.scss:
```diff
@@ -0,0 +1,183 @@
+        margin-bottom: ($baseline*0.75);
+
+        .u-field-value {
+            font-size: 1.7em;
```
use %t-title4 here instead of font-size and line-height

frrrances commented on lms/static/sass/views/_learner-profile.scss:
```diff
@@ -0,0 +1,183 @@
+    }
+
+    .u-field-message {
+        float: right;
```
support rtl with @include float(right);

frrrances commented on lms/static/sass/views/_learner-profile.scss:
```diff
@@ -0,0 +1,183 @@
+    }
+
+    .u-field-value {
+        width: 200px;
```
this is cutting off longer usernames. maybe add a wider width to the rule at line 103.

frrrances commented on lms/static/sass/views/_learner-profile.scss:
```diff
@@ -0,0 +1,183 @@
+
+    .u-field-title {
+        font-size: 1.1em;
+        font-weight: $font-bold;
```
use %t-weight4

frrrances commented on lms/static/sass/views/_learner-profile.scss:
```diff
@@ -0,0 +1,183 @@
+
+    .u-field.mode-placeholder {
+        padding: $baseline;
+        border: 1px dashed $base-font-color;
```
make this 2px and use color $link-color

frrrances commented on lms/static/sass/views/_learner-profile.scss:
```diff
@@ -0,0 +1,183 @@
+    }
+
+    .u-field.mode-placeholder:hover {
+        border: 1px dashed $m-blue-d1;
```
make this 2px and use $link-color

frrrances commented on lms/static/sass/views/_learner-profile.scss:
```diff
@@ -0,0 +1,183 @@
+    .u-field.mode-placeholder:hover {
+        border: 1px dashed $m-blue-d1;
+        .u-field-title {
+          color: $m-blue-d1;
```
use $link-color

frrrances commented:
@muzaffaryousaf Finished my code review. Let me know when you have the changes up and I will review again!

muzaffaryousaf commented:
@frrrances i've addressed all of your comments except the background colour for this page. I've added that to my page's wrapper element but i think it's not a better way to do this. Can you please suggest me how i can set that on `content-wrapper` class ? Also, please let me know if i'm missing something.

andy-armstrong commented:
@muzaffaryousaf Can you rebase this as currently it is showing all the commits from learner profiles too.

marcotuts commented:
@muzaffaryousaf - I'm seeing an issue on the sandbox where I can still toggle to the Full Profile even if I've not set my age in Account Settings. I don't think this should be possible, and perhaps this is a part of a future story, but I wanted to double check. (cc'ed @explorerleslie )

andy-armstrong commented:
@muzaffaryousaf There's a small bug where the drop down in the top right shows Account Settings as selected when you are actually on the profile page.

![image](https://cloud.githubusercontent.com/assets/5985072/6846698/95f60a00-d396-11e4-8f0d-964fee58e747.png)

andy-armstrong commented:
@muzaffaryousaf There is a small bug where the menu in the top right shows Account Settings as the selected page even when you are on the profile page.

![image](https://cloud.githubusercontent.com/assets/5985072/6846717/c1d8e4c6-d396-11e4-9052-3fbb221a4a90.png)

frrrances commented on lms/templates/student_profile/learner_profile.html:
```diff
@@ -0,0 +1,41 @@
+<%namespace name='static' file='/static_content.html'/>
+
+<%block name="pagetitle">${_("Learner Profile")}</%block>
+
```
Add body class here: 

```
<%block name="bodyclass">view-profile</%block>
```

frrrances commented on lms/static/sass/views/_learner-profile.scss:
```diff
@@ -0,0 +1,188 @@
+// * +Container - Learner Profile
+// * +Main - Header
+// * +Settings Section
+
```
add wrapper class for this whole sass partial: 

```
.view-profile {
...
}
```

frrrances commented on lms/static/sass/views/_learner-profile.scss:
```diff
@@ -0,0 +1,188 @@
+// * +Settings Section
+
+$profile-photo-dimension: 120px;
+
```
add this rule to make the content background white: 

```
.content-wrapper {
  background-color: $white;
}
```

frrrances commented on lms/static/sass/views/_learner-profile.scss:
```diff
@@ -0,0 +1,188 @@
+
+.wrapper-profile {
+    min-height: 200px;
+    background-color: $white;
```
you can remove this background color.

frrrances commented:
@muzaffaryousaf Thanks for making those changes. I added a few more comments that should help you scope the rules to just the learner profile by adding a class to the body, then adding the white background to the .view-profile .content-wrapper {}

frrrances commented on lms/static/sass/shared/_fields.scss:
```diff
@@ -0,0 +1,103 @@
+.u-field {
+    padding: $baseline 0;
+    border-bottom: 1px solid $gray-l5;
+    border: 1px dashed transparent;
```
Some of these rules apply only to the profile view so I think they should move into that sass file. Here is the edited set of rules: 

shared/_fields.scss: 
```
.u-field {
  padding: $baseline 0;
  border-bottom: 1px solid $gray-l5;
}
```

views/_learner-profile.scss: 
```
.u-field {
  border: 2px dashed transparent;
  border-radius: 3px;

  &.mode-placeholder:hover {
    border: 2px dashed $link-color;
    }  

  &.editable-toggle.mode-display:hover {
    background-color: $m-blue-l4;
    border-radius: 3px;

    .message-can-edit {
      display: inline-block;
      color: $link-color;
    }
  }
}

```

muzaffaryousaf commented:
@frrrances Thanks for review. Also, i've addressed all of the comments. Please let me know if i'm missing something. Thanks.

muzaffaryousaf commented:
@andy-armstrong i've rebased the branch.

andy-armstrong commented on common/test/acceptance/pages/lms/learner_profile.py:
```diff
@@ -0,0 +1,147 @@
+    """
+
+    def __init__(self, browser, username):
+        """
```
I think that __init__ is ignored by PyLint for missing docstrings.

andy-armstrong commented on common/test/acceptance/pages/lms/learner_profile.py:
```diff
@@ -0,0 +1,147 @@
+        """
+        self.value_for_dropdown_field('language', value)
+
+    def aboutme(self, value=None):
```
I'd call this "bio" for consistency.

andy-armstrong replied:
Never mind. I see that the UX calls this "About Me" now. Nit: I'd probably call it ``about_me``.

andy-armstrong commented on common/test/acceptance/pages/lms/learner_profile.py:
```diff
@@ -0,0 +1,147 @@
+        self.wait_for_ajax()
+        return self.q(css='#u-field-select-account_privacy').visible
+
+    def country(self, value=None):
```
Can these three fields also be ``@property`` or does that not work when they have kwargs?

andy-armstrong commented on common/test/acceptance/tests/lms/test_learner_profile.py:
```diff
@@ -0,0 +1,259 @@
+        And I reload the page.
+        Then I should see the profile visibility selector dropdown.
+        Then I see all the profile fields are shown.
+        And `location`, `language` and `about me` fields are editable.
```
Please add tests for setting each of these fields that verify that the values are persisted after reloading the page.

andy-armstrong replied:
Never mind. I didn't notice that you had those tests below because they didn't have the BDD specs.

frrrances commented on lms/static/sass/views/_learner-profile.scss:
```diff
@@ -0,0 +1,193 @@
+
+        .u-field-value {
+            width: 100%;
+            white-space: pre-line;
```
add line-height: 1.5;

andy-armstrong commented on common/test/acceptance/tests/lms/test_learner_profile.py:
```diff
@@ -0,0 +1,259 @@
+
+    def test_country_field(self):
+        """
+        Test behaviour of `Country` field.
```
Please make these comments into BDD-style specs like the ones above. I thought these were helper methods initially. I'd also reload the page in each instance and verify that the value has persisted.

frrrances commented on lms/static/sass/views/_learner-profile.scss:
```diff
@@ -0,0 +1,193 @@
+            i {
+              font-size: 12px;
+              padding-right: 5px;
+              vertical-align: middle;
```
add color: $gray-l1;

frrrances commented on lms/static/sass/views/_learner-profile.scss:
```diff
@@ -0,0 +1,193 @@
+
+            .u-field-value {
+                text-align: center;
+                line-height: 1.3em;
```
change line-height to 1.5 and add @extend %t-copy-sub1; color: $gray-d1;

andy-armstrong commented on lms/static/js/spec/student_profile/learner_profile_view_spec.js:
```diff
@@ -0,0 +1,176 @@
+                Helpers.expectLoadingErrorIsVisible(learnerProfileView, false);
+                Helpers.expectLimitedProfileSectionsAndFieldsToBeRendered(learnerProfileView, true);
+            });
+        });
```
Please add tests for editing each of the editable fields with both successful responses and for error responses.

frrrances commented on lms/static/sass/shared/_fields.scss:
```diff
@@ -33,10 +75,14 @@
     width: flex-grid(3, 12);
     display: inline-block;
-    color: $dark-gray1;
+    color: $gray-l1;
```
Sorry it needs to be changed to $gray-d1 for accessibility.

andy-armstrong commented:
:+1: This looks great to me. I have a few comments about testing and then this looks ready to go.

frrrances commented on lms/static/sass/shared/_fields.scss:
```diff
@@ -33,10 +75,14 @@
 }
 
+.u-field-icon {
+    width: $baseline;
```
add color: $gray-l2;

frrrances commented on lms/static/sass/shared/_fields.scss:
```diff
@@ -5,6 +5,48 @@
+        border-radius: 3px;
+
+        span {
+            color: $gray-l3;
```
Sorry please change to $gray-l1;

frrrances commented on lms/static/sass/shared/_fields.scss:
```diff
@@ -5,6 +5,48 @@
+    }
+
+    i {
+      color: $gray;
```
Change to $gray-l2. Also add margin-right: 5px;

frrrances commented:
I couldn't find it in the code, but please remove the "fa-flip-horizontal" on the language icon. Other than that, I added a few other small things and once you address those, I think this is ready! :+1:

muzaffaryousaf commented:
@cptvitamin can you please suggest about the focus issue that @andy-armstrong reported. It's happening because we set the focus on first `li` item in any case. One way is to only remove that focus but i'm not sure how it can affect the accessibility of this feature.

>There is a small bug where the menu in the top right shows Account Settings as the selected page even when you are on the profile page.
![image](https://cloud.githubusercontent.com/assets/5985072/6846717/c1d8e4c6-d396-11e4-9052-3fbb221a4a90.png)

muzaffaryousaf commented:
@explorerleslie can you please suggest about these points.

By @marcotuts :
> I'm seeing an issue on the sandbox where I can still toggle to the Full Profile even if I've not set my age in Account Settings. I don't think this should be possible, and perhaps this is a part of a future story, but I wanted to double check. (cc'ed @explorerleslie )


By @srpearce :
> Text: This edX learner is currently sharing a limited profile.
Comment: Should this message be different for the user herself and for other page viewers? It's a little strange for me to see "This edX learner is not currently sharing their profile details" on my own profile page. If we don't want to have different messages, can we indicate somewhere on the UI that other users will see the profile page exactly as it looks here - maybe a "Other edX learners will see your profile exactly as it appears below" under the "edX learners can see my" line? (Maybe this is a follow-on story?)
Another potential follow-on story: Let users know that if they make changes to the full profile, those changes won't be deleted if the user switches to the limited profile and then back to the full profile. It could be a message like "This information will be saved but not visible when you switch to a limited profile" (though said much more elegantly than this) above the "About Me" text box. (I would add this above the text box rather than as placeholder text so that it's always visible.)

muzaffaryousaf commented:
test this please

srpearce commented:
Responding to @muzaffaryousaf's comment above:

If you're looking at your own profile: You are currently sharing a limited profile.

If you're looking at someone else's profile:  This edX learner is currently sharing a limited profile.

You can ignore the suggestion about a follow-on story for now.

cptvitamin commented:
@andy-armstrong, @muzaffaryousaf is right, the ring around the Account Settings option in the menu does not indicate what page is active, just what menu item has focus.  As recommended in [WAI-ARIA Best Practices](http://www.w3.org/TR/wai-aria-practices/#menu) when activating a menu, focus should move to the first item in the list of menu items, which is how we designed that menu.  So this is not a bug.  If you want a separate style to indicate the page that is active, we would have to define a new style for that.  "Menus" like this are supposed to mimic menus that we find in the desktop environment, which do not indicate the current page/view, so I personally don't think it's necessary.

cptvitamin commented on lms/static/certificates/sass/main-ltr.css:
```diff
@@ -1010,667 +1010,7 @@ abbr[title] {
-      .view-validate-certificate .support .list-actions .action {
-        display: block;
-        text-align: center; } }
+.header-app{max-width:1280px;margin-left:auto;margin-right:auto}.header-app::after{clear:both;content:"";display:table}.title-logo{text-align:center}.title-logo .logo{display:inline-block;vertical-align:middle;margin-right:20px}.title-logo .title-sub{display:inline-block;vertical-align:middle;text-transform:uppercase;letter-spacing:0.1rem}.wrapper-footer{padding:0 40px;background:transparent;colo ...
```
Is there a particular reason why all this CSS got minimized here?

muzaffaryousaf replied:
this is being done by  mistake. i'm going to revert these changes.

srpearce commented on lms/static/js/spec/student_account/helpers.js:
```diff
@@ -85,6 +86,92 @@ define(['underscore'], function(_) {
+            expectProfileElementContainsField(sectionFieldElement, learnerProfileView.options.sectionOneFieldViews[fieldIndex]);
+        });
+
+        expect($('.profile-private--message').text()).toBe('This edX learner is not currently sharing their profile details.')
```
If you're looking at your own profile: You are currently sharing a limited profile.

If you're looking at someone else's profile: This edX learner is currently sharing a limited profile.

cptvitamin commented on lms/static/certificates/sass/main-rtl.css:
```diff
@@ -1010,667 +1010,7 @@ abbr[title] {
-      .view-validate-certificate .support .list-actions .action {
-        display: block;
-        text-align: center; } }
+.header-app{max-width:1280px;margin-left:auto;margin-right:auto}.header-app::after{clear:both;content:"";display:table}.title-logo{text-align:center}.title-logo .logo{display:inline-block;vertical-align:middle;margin-left:20px}.title-logo .title-sub{display:inline-block;vertical-align:middle;text-transform:uppercase;letter-spacing:0.1rem}.wrapper-footer{padding:0 40px;background:transparent;color ...
```
Is there a particular reason why all this CSS got minimized here?

cptvitamin commented on lms/static/js/views/fields.js:
```diff
@@ -20,13 +20,16 @@
             tagName: 'div',
 
             indicators: {
+                'canEdit': '<i class="icon fa fa-pencil message-can-edit"></i>',
```
all of these icon fonts should have ```aria-hidden="true"``` attributes on them.  If they appear by themselves, without text, then screen reader only text must be added that indicates the purpose of the icon i.e. ```<span class="sr">Editable</span>```

cptvitamin commented on lms/templates/fields/field_dropdown.underscore:
```diff
@@ -1,16 +1,26 @@
+<% } %>
+
+<% if (iconName) { %>
+    <i class="u-field-icon icon fa <%- iconName %> fa-fw"></i>
```
icon fonts need to have ```aria-hidden="true"``` attributes on them.

srpearce commented on lms/static/js/student_profile/views/learner_profile_fields.js:
```diff
@@ -0,0 +1,38 @@
+            message: function () {
+                if (this.profileIsPrivate) {
+                    this._super(interpolate_text(
+                        gettext("Your profile is disabled because you haven't filled in your Year of Birth. Please visit the Account Settings page. {account_settings_page_link}."),
```
You must specify your birth year before you can share your full profile. To specify your birth year, go to the Account Settings page. (make "Account Settings page" a link)

srpearce commented on lms/static/js/student_profile/views/learner_profile_fields.js:
```diff
@@ -0,0 +1,38 @@
+                    ));
+                } else if (this.requiresParentalConsent) {
+                    this._super(interpolate_text(
+                        gettext('Your profile is disabled because you are under 14. If this is incorrect, please visit the Account Settings page. {account_settings_page_link}.'),
```
You must be over 13 to share a full profile. If you are over 13, make sure that you have specified a birth year on the Account Settings page. (make "Account Settings page" a link)

srpearce commented on lms/templates/student_profile/learner_profile.underscore:
```diff
@@ -0,0 +1,25 @@
+        <div class="wrapper-profile-section-two">
+            <div class="profile-section-two-fields">
+                <% if (!showFullProfile) { %>
+                    <span class="profile-private--message"><%- gettext('This edX learner is not currently sharing their profile details.') %></span>
```
If you're looking at your own profile: You are currently sharing a limited profile.

If you're looking at someone else's profile: This edX learner is currently sharing a limited profile.

cptvitamin commented:
This is looking good @muzaffaryousaf.  One comment I have is that the contrast of the labels and the help text is really low.  To meet minimum contrast requirements, we will have to go with rgb(118, 118, 118) or #767676 or darker.  cc @frrrances

andy-armstrong commented:
@cptvitamin Thanks for the explanation. I read it as being the current page since we do that in Studio. If that's not the intention then the current behavior makes sense.

![image](https://cloud.githubusercontent.com/assets/5985072/6874596/d78429ae-d48e-11e4-9627-f46999f77071.png)

muzaffaryousaf commented:
@srpearce i've addressed all of your suggestions. can you please give it a final look. Thanks.

cptvitamin commented:
@muzaffaryousaf, there seems to be a serious bug on the "My Profile" page.  I cannot tab beyond the "Learners can see..." combo box.  Subsequent tabs move focus (as far as I can tell) just before the label element for that combo box.  This means a screen reader or keyboard only user can never access anything beyond that on the page.  It seems to be that the JS is overriding the default tab order.

cptvitamin commented on lms/templates/student_profile/learner_profile.underscore:
```diff
@@ -0,0 +1,29 @@
+    <div class="wrapper-profile-sections account-settings-container">
+        <div class="wrapper-profile-section-one">
+            <div class="profile-photo">
+                <img src="<%- profilePhoto %>">
```
@muzaffaryousaf, this img must have an alt attribute.  I recommend ```alt="Profile image for [USERNAME]"```

explorerleslie commented:
> By @marcotuts :
> I'm seeing an issue on the sandbox where I can still toggle to the Full Profile even if I've not set my age in Account Settings. I don't think this should be possible, and perhaps this is a part of a future story, but I wanted to double check. (cc'ed @explorerleslie )

@muzaffaryousaf @marcotuts this is part of https://openedx.atlassian.net/browse/TNL-1658

explorerleslie commented:
From Sylvia: 
> Responding to @muzaffaryousaf's comment above:
> If you're looking at your own profile: You are currently sharing a limited profile.
> If you're looking at someone else's profile: This edX learner is currently sharing a limited profile.
> You can ignore the suggestion about a follow-on story for now.

@muzaffaryousaf what @srpearce said. :)

srpearce commented on lms/static/js/student_profile/views/learner_profile_fields.js:
```diff
@@ -0,0 +1,38 @@
+                if (this.profileIsPrivate) {
+                    this._super(interpolate_text(
+                        gettext("You must specify your birth year before you can share your full profile. To specify your birth year, go to the {account_settings_page_link}"),
+                        {'account_settings_page_link': '<a href="' + this.options.accountSettingsPageUrl + '">' + gettext('Account Settings page,') + '</a>'}
```
The comma after "Account Settings page" should be a period.

srpearce commented on lms/templates/student_profile/learner_profile.underscore:
```diff
@@ -0,0 +1,29 @@
+            <div class="profile-section-two-fields">
+                <% if (!showFullProfile) { %>
+                    <% if(ownProfile) { %>
+                        <span class="profile-private--message"><%- gettext("If you're looking at your own profile: You are currently sharing a limited profile.") %></span>
```
@muzaffaryousaf Sorry, the part before the colon was supposed to explain where this goes, not be part of the UI text. The text should just be:

You are currently sharing a limited profile.

muzaffaryousaf replied:
:) sorry my mistake.. I should have seen this earlier :(

srpearce commented on lms/templates/student_profile/learner_profile.underscore:
```diff
@@ -0,0 +1,29 @@
+                    <% if(ownProfile) { %>
+                        <span class="profile-private--message"><%- gettext("If you're looking at your own profile: You are currently sharing a limited profile.") %></span>
+                    <% } else {  %>
+                        <span class="profile-private--message"><%- gettext("If you're looking at someone else's profile: This edX learner is currently sharing a limited profile.") %></span>
```
@muzaffaryousaf Sorry, the part before the colon was supposed to explain where this goes, not be part of the UI text. The text should just be:

This edX learner is currently sharing a limited profile.

cptvitamin commented:
@muzaffaryousaf, I'm starting to find some major accessibility problems on the My Profile page.  The custom widgets to edit the location, language and bio are not accessible at all.  The short story is they are not focusable and provide no role, state or property information.  

Here is the long story:
* When you use elements with no implied semantics or behaviors, like divs and spans and use them as dynamic elements, scripting their behavior, you have to make sure you script *all* of the behavior, including roles, states and properties.  Since you are actually using native form elements to actually make the changes, this won't be too hard.
* The Location, Language and Bio ```<spans>``` need to have a ```tabindex="0"``` to make them keyboard focusable.
* The Location, Language and Bio ```<spans>``` need to identify themselves as the actual input field types they will be converted to on focus.  So the Location and Language ```<spans>``` need to have ```role="listbox"``` and ```aria-expanded="false"```.  It also needs to have an ```aria-label="Location"``` or ```Language``` http://www.w3.org/TR/wai-aria/roles#listbox 
* The Bio div needs to have an ```aria-role="textbox"```.  I think using aria-labelledby to reference the text "About Me" is probably best (you will have to add an ID to it) and then use aria-describedby to reference the text "Tell other edX learners a little about yourself: where you live, what your interests are, why you're taking courses on edX, or what you hope to learn."  Again, you will need to ref it by ID.
* You will want to make sure that as the DOM elements are swapped that focus is logically maintained.
* You will also want to add aria-live regions to any any element that is expected to receive dynamic notification text.  The trick is that these must be in the DOM before the text is injected, you can't inject an aria-live region.  I noticed this issue when I received an error message regarding the textarea.  The code would look something like this:
```
<span aria-live="polite"></span>
```
Then on error:
```
<span aria-live="polite">
   <i class="fa fa-exclamation-triangle message-error" aria-hidden="true"></i>
   <span class="sr">Explanation of the error</span>
</span>
```

This will read the message to a screen reader no matter where their focus is at the time.

You can create a separate PR for this work, but it will need to be fixed before this goes lives.  It will most likely need one or two more rounds of testing to make sure we get it right.  There is also the case of that strange bug that prevents moving focus beyond the dropdown menu...

You may find it useful to read through the Accessibility Developers Guide we recently put together: http://edx.readthedocs.org/projects/edx-developer-guide/en/latest/accessibility.html

cc @explorerleslie @marcotuts @frrrances

srpearce commented:
@muzaffaryousaf Just a few small corrections and we're good.

muzaffaryousaf commented:
>@muzaffaryousaf, I'm starting to find some major accessibility problems on the My Profile page. The custom widgets to edit the location, language and bio are not accessible at all. The short story is they are not focusable and provide no role, state or property information.

@cptvitamin thanks for review, but i would be doing these changes in separate PR and will tag you there, as we are about to merge this.

muzaffaryousaf commented:
@symbolist i've addressed all the comments and tests are passing now. Can you give it a final look.
Thanks.

symbolist commented:
Looks good to go to me! :+1:

andy-armstrong commented:
@muzaffaryousaf @symbolist @explorerleslie - @cahrens just reminded me of one issue that we'd discussed but which I hadn't mentioned to you. It is a little strange to have a profile image show a loading indicator and then have to wait for an AJAX call, when the only information shown on the page appears to be static to the end user. One trick we've used on other pages is to compute the JSON for the model on the server and to inline it into the page. This means that the page downloads with everything it needs, so it doesn't need a subsequent AJAX call to render itself.

This is not a big deal for now, but we should consider implementing it this way if you need to go back into this code again. It is a very simple change that has quite a nice UX improvement for the typical user. Note that this isn't noticeable on the sandboxes because the AJAX call is so fast, but in production the loading indicator could take a few seconds to go away.
